backend/src/routes/users.js

```diff
@@ -2,577 +2,353 @@
-      check('permissions.personal_image').isBoolean(),
-      check('permissions.email_list').isBoolean(),
-      check('permissions.signature')
-        .isAscii()
-        .trim()
-        .escape(),
-      check('events').exists()
-    ]),
-    (req, res, query) => {
+    oneOf(USER_DATA_VALIDATIONS),
+    (req, res, next) => {
       const errors = validationResult(req);
       if (!errors.isEmpty()) {
         return res.status(400).json({ errors: errors.mapped() });
       }
 
-      const userData = matchedData(req);
-      userData.events = req.body.events;
+      const userDataReq = matchedData(req);
+      userDataReq.events = req.body.events;
 
-      if (userData.bio != undefined && userData.bio.password) {
-        // if we want to update password
-        bcrypt
-          .hash(userData.bio.password, 10)
-          .then(hash => {
-            userData.bio.password = hash;
-            return User.findById(req.params.id);
-          })
-          .then(user => {
-            if (!user) {
-              return res.status(404).json({ errors: `No user found with id: ${req.params.id}` });
-            }
-            userData.id = undefined; // we do not want to update the user's id
-            for (const key1 in userData) {
-              if (userData.hasOwnProperty(key1)) {
-                const obj = userData[key1];
-                const userObj = user[key1];
-                for (const key2 in obj) {
-                  if (obj.hasOwnProperty(key2)) {
-                    userObj[key2] = obj[key2] !== undefined ? obj[key2] : userObj[key2];
-                  }
-                }
-                user[key1] = userObj;
-              }
-            }
-
-            user.save();
-            return res.status(200).json({ user });
-          })
-          .catch(errors => {
-            console.log(errors);
-            res.status(500).json({ errors });
-          });
-      } else {
-        // we do not want to update password
-        User.findById(req.params.id)
-          .then(user => {
-            if (!user) {
-              return res.status(404).json({ errors: `No user found with id: ${req.params.id}` });
-            } else if (req.query.action === 'appendEvent') {
-              userData.events.forEach(eventId => user.events.push(eventId));
-              userData.events = undefined;
-            } else if (req.query.action === 'removeEvents') {
-              userData.events.forEach(eventId =>
-                user.events.splice(user.events.indexOf(eventId), 1)
-              );
-              userData.events = undefined;
-            }
+      UserData.findById(req.params.id)
+        .then(user => {
+          if (!user) {
+            return res.status(404).json({ errors: `No user found with id: ${req.params.id}` });
+          } else if (req.query.action === 'appendEvent') {
+            userDataReq.events.forEach(eventId => user.events.push(eventId));
+            userDataReq.events = undefined;
```

Comment:
It is typically considered a bad idea to set things to `undefined`. There's varying schools of thought on the matter, but in general the idea is that you should just let JS set things to `undefined` when they don't have a value. If you are in a situation where you want to set something to `undefined`, it's typically better to either set it to `null` or use the `delete key` operator to remove it from the object.

Reply:
Ah ok, that makes sense. Will use `delete` instead.